### PR TITLE
[SPARK-45593][BUILD] Building a runnable distribution from master code running spark-sql raise error

### PR DESCRIPTION
**Pull request by yikf**
### What changes were proposed in this pull request?

Fix a build issue, when building a runnable distribution from master code running spark-sql raise error:
```
Caused by: java.lang.ClassNotFoundException: org.sparkproject.guava.util.concurrent.internal.InternalFutureFailureAccess
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 58 more
```
the problem is due to a gauva dependency in  spark-connect-common POM that **conflicts**  with the shade plugin of the parent pom.

- the spark-connect-common contains `connect.guava.version` version of guava, and it is relocation as `${spark.shade.packageName}.guava` not the `${spark.shade.packageName}.connect.guava`;
- The spark-network-common also contains guava related classes, it has also been relocation is `${spark.shade.packageName}.guava`, but guava version `${guava.version}`;
- As a result, in the presence of different versions of the classpath org.sparkproject.guava.xx;

In addition, after investigation, it seems that module spark-connect-common is not related to guava, so we can remove guava dependency from spark-connect-common.

### Why are the changes needed?

Building a runnable distribution from master code is not runnable.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

I ran the build command output a runnable distribution package manually for the tests; 

Build command:
```
./dev/make-distribution.sh --name ui --pip --tgz  -Phive -Phive-thriftserver -Pyarn -Pconnect
```

Test result:
image

I also checked the `org.sparkproject.guava.cache.LocalCache` from jars dir;
Before:
```
➜  jars grep -lr 'org.sparkproject.guava.cache.LocalCache' ./
.//spark-connect_2.13-4.0.0-SNAPSHOT.jar
.//spark-network-common_2.13-4.0.0-SNAPSHOT.jar
.//spark-connect-common_2.13-4.0.0-SNAPSHOT.jar
```

Now:
```
➜  jars grep -lr 'org.sparkproject.guava.cache.LocalCache' ./
.//spark-network-common_2.13-4.0.0-SNAPSHOT.jar
```

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by Yikf** [comment]
Kindly ping @juliuszsompolski @LuciferYang @dongjoon-hyun , Could you please take a look if you find a moment ~

**Comment by LuciferYang** [comment]
hmm... also cc @hvanhovell

**Comment by LuciferYang** [comment]
> Kindly ping @juliuszsompolski @LuciferYang @dongjoon-hyun , Could you please take a look if you find a moment ~

I will discuss the specifics with you offline tomorrow.

File: connector/connect/common/pom.xml [hunk 1]
```diff
@@ -47,18 +47,6 @@
             <groupId>com.google.protobuf</groupId>
             <artifactId>protobuf-java</artifactId>
         </dependency>
-        <dependency>
```
[the remark this answers is not in the document]
Comment (Yikf) [reply connector/connect/common/pom.xml:50]: Both connect-server and connect-client  have their own independent guava dependencies : )

I think we should use connect-common shade guava, connect-client and connect-server shade connect-common to make connect-server and connect-client use the same guava dependency. 

In addition, because connect-common is included in the connect-server shade, the runnable distribution can also exclude connect-common. 

Do you like this current commit?

[the remark this answers is not in the document]
Comment (Yikf) [reply connector/connect/common/pom.xml]: I tested it a s a's description and it worked as scheduled
I tested as https://github.com/apache/spark/pull/43195#issue-1921234067, i 
image

**Comment by LuciferYang** [comment]
A bit busy today, I'll check this PR later.

File: assembly/pom.xml [hunk 2]
```diff
@@ -149,6 +149,12 @@
           <groupId>org.apache.spark</groupId>
           <artifactId>spark-connect_${scala.binary.version}</artifactId>
           <version>${project.version}</version>
+          <exclusions>
+            <exclusion>
+              <groupId>org.apache.spark</groupId>
+              <artifactId>spark-connect-common_${scala.binary.version}</artifactId>
```
Comment (hvanhovell) [inline assembly/pom.xml:155]: Why this change?

File: connector/connect/client/jvm/pom.xml [hunk 3]
```diff
@@ -124,6 +124,10 @@
                 <include>io.grpc.**</include>
               </includes>
             </relocation>
+            <relocation>
```
Comment (hvanhovell) [inline connector/connect/client/jvm/pom.xml:127]: So in how many places will we shade Guava now? 2? or 3? AFAICT we have the relocation defined by the parent, this one, and the one below that might come into play?

File: connector/connect/common/pom.xml [hunk 4]
```diff
@@ -140,6 +140,35 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
```
Comment (hvanhovell) [inline connector/connect/common/pom.xml:143]: Why do we need to shade common? For the off-chance someone takes a direct dependency on it?

File: assembly/pom.xml (hunk 2 above)
Comment (Yikf) [reply assembly/pom.xml:155]: Runnable distribution tar already contains connect-server and connect-server shade the connect-common, so runnable distribution tar does not need to contain connect-common anymore.
https://github.com/apache/spark/blob/8cdcfd262f9fd46fb9a8e1ceb0bccefe452582bd/connector/connect/server/pom.xml#L306

File: connector/connect/client/jvm/pom.xml (hunk 3 above)
Comment (Yikf) [reply connector/connect/client/jvm/pom.xml:127]: The current Apache Spark includes two versions of Guava, which are (The historical reasons are not clear to me. @LuciferYang may know some background):
1. https://github.com/apache/spark/blob/8cdcfd262f9fd46fb9a8e1ceb0bccefe452582bd/pom.xml#L203
2. https://github.com/apache/spark/blob/8cdcfd262f9fd46fb9a8e1ceb0bccefe452582bd/pom.xml#L292

The current Apache Spark has shading in the following three places for these two versions:
1. for guava.version
https://github.com/apache/spark/blob/8cdcfd262f9fd46fb9a8e1ceb0bccefe452582bd/common/network-common/pom.xml#L125-L129
2. for connect.guava.version
https://github.com/apache/spark/blob/8cdcfd262f9fd46fb9a8e1ceb0bccefe452582bd/connector/connect/server/pom.xml#L310-L316
3. for conflicting versions, the connect-common does not override the parent's shade plugin, so it is relocated to '${spark.shade.packageName}.guava'. However, the Guava version of connect-common is connect.guava.version, which results in two versions of the same package in the classpath.
https://github.com/apache/spark/blob/8cdcfd262f9fd46fb9a8e1ceb0bccefe452582bd/connector/connect/common/pom.xml#L56-L61

This PR plans to make modifications following this line of thought,  we should use connect-common shade guava, connect-client and connect-server shade connect-common to make connect-server and connect-client use the same guava dependency.

Comment (LuciferYang) [reply connector/connect/client/jvm/pom.xml:127]: > The current Apache Spark includes two versions of Guava, which are (The historical reasons are not clear to me. @LuciferYang may know some background):

Sorry I missed this ping. 14.0.1 is because Hive 2.3.9 depends on Guava 14.0.1, which is shaded in the `network-common` module and used by other non-connect modules. `32.0.1-jre` is dedicated to Connect-related modules. Currently, the connect server and connect jvm client modules should have shaded it respectively, and for the Guava 32.0.1-jre, the dependency `com.google.guava:failureaccess` should also be shaded.

Comment (LuciferYang) [reply connector/connect/client/jvm/pom.xml:127]: From the error message, it seems that the connect-common module has inherited the shade plugin configuration from the parent pom.xml, and then performed shade + relocation on Guava 32.0.1-jre, but missed `com.google.guava:failureaccess`. Perhaps overwriting the configuration of the shade plugin in the connect-common module to do nothing would solve the problem?

Comment (Yikf) [reply connector/connect/client/jvm/pom.xml:127]: yea

Comment (c3-ffomenko) [reply connector/connect/client/jvm/pom.xml:127]: Thank you for working on this PR, it helped us with our issue. We had additional problem with same error message when we build spark with upgraded version of guava which depends on failureaccess. In this case shading done in network common misses failureaccess dependency since shading is done only on guava omitting classes in failureaccess.

**Comment by Yikf** [comment]
Any other suggestion on the PR? @hvanhovell @LuciferYang

File: connector/connect/client/jvm/pom.xml [hunk 5]
```diff
@@ -137,6 +137,10 @@
...
```
[the remark this answers is not in the document]
Comment (Yikf) [reply connector/connect/client/jvm/pom.xml]: you are right, It is unnecessary. 

Now, we perform shading of the Guava library within the connect-common module to ensure both connect-server and connect-client modules maintain consistent and accurate Guava dependencies.

[the remark this answers is not in the document]
Comment (Yikf) [reply connector/connect/client/jvm/pom.xml]: BTW, I also tested connect as described in [PR description](https://github.com/apache/spark/pull/43195#issue-1921234067) , and spark sql shell using `./bin/spark-sql` , and it worked as expected.

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Comment by LuciferYang** [comment]
@Yikf Can you re-trigger the failed GA tasks?

@hvanhovell Do you have any further suggestions for this pr? If not, I will merge it after it passes all GA tests. Thanks

**Comment by LuciferYang** [comment]
Merged into master. Thanks @Yikf @hvanhovell @c3-ffomenko
